[c10d] Error out the case when registering symmetric memory without eager init

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160145

Instead of implicitly creating nccl comm inside mem pool registration for symmetric memory, we decide to error it out so that we only support eager init case when the nccl comm is already initiated.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta

fduwjj committed:
[c10d] Error out the case when registering symmetric memory without eager init

[ghstack-poisoned]

syed-ahmed commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1098,6 +1098,11 @@ void ProcessGroupNCCL::registerMemPool(c10::cuda::MemPool* pool) {
+    }
     // HACK: currently we are using this function for NVLS
     // reductions, and that's why using OpType::ALLREDUCE.
     // If we end up using this API for zero-copy P2P, we might
```
We should remove this note and initNCCLComm below?

fduwjj replied:
oh really, I see.

fduwjj committed:
Update on "[c10d] Error out the case when registering symmetric memory without eager init"


Instead of implicitly creating nccl comm inside mem pool registration for symmetric memory, we decide to error it out so that we only support eager init case when the nccl comm is already initiated.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

Skylion007 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3172,11 +3172,16 @@ def test_nccl_user_buffer_registration(self):
     @requires_multicast_support()
     def test_nccl_window_registration(self):
         store = c10d.FileStore(self.file_name, self.world_size)
+        torch.cuda.set_device(self.rank)
```
Ugh, should we need put this in a with context and return the default cuda device to original rank? Otherwise order of ops could matter right?

ngimel replied:
yes, should put it in context manager

fduwjj committed:
Update on "[c10d] Error out the case when registering symmetric memory without eager init"


Instead of implicitly creating nccl comm inside mem pool registration for symmetric memory, we decide to error it out so that we only support eager init case when the nccl comm is already initiated.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

kwen2501 reviewed (approved):
LGTM.
cc @syed-ahmed in case there are other thoughts. [Edit: just saw that @syed-ahmed has reviewed this PR too]

fduwjj committed:
Update on "[c10d] Error out the case when registering symmetric memory without eager init"


Instead of implicitly creating nccl comm inside mem pool registration for symmetric memory, we decide to error it out so that we only support eager init case when the nccl comm is already initiated.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

ngimel commented:
I'm slightly surprised you didn't need to change test_nccl_user_buffer_registration
https://github.com/pytorch/pytorch/blob/e46458196880d4091cb38dc3093fbfc21c6381ff/test/distributed/test_c10d_nccl.py#L3130-L3146, is it because we are using nccl comms between tests and by the time we run it communicator is already created? Would this test fail if run standalone?
